### ISSUE TITLE
[Bug] The `context` keyword is available when looking for references, not definitions

### ISSUE THREAD

**Issue by pidgeon777** [ray-x/navigator.lua#164]
In my opinion, this line should be removed:

https://github.com/ray-x/navigator.lua/blob/d990da84a6c267bd8684535d392e551333d7a919/lua/navigator/reference.lua#L153

When looking for references, the attached LSP server receives a request for definition search by also specifying the `IncludeDeclaration = false` argument.

This is supported when looking for references, but not definitions:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.referencecontext?view=visualstudiosdk-2022

My LSP server returns this error when looking for references of a symbol:

```log
2022-04-05 11:32:48,123 [DEBUG] Read msg: {'params': {'position': {'character': 14, 'line': 667}, 'context': {'includeDeclaration': False}, 'textDocument': {'uri': 'file:///C:/I2C_MASTER.vhd'}}, 'jsonrpc': '2.0', 'method': 'textDocument/definition', 'id': 2}
2022-04-05 11:32:48,123 [ERROR] Caught exception while handling textDocument/definition with params {'position': {'character': 14, 'line': 667}, 'context': {'includeDeclaration': False}, 'textDocument': {'uri': 'file:///C:/I2C_MASTER.vhd'}}:
Traceback (most recent call last):
  File "C:\msys64\mingw64\lib\python3.9\site-packages\pyGHDL\lsp\lsp.py", line 122, in handle
    response = fmethod(**params)
TypeError: textDocument_definition() got an unexpected keyword argument 'context'
```

**Comment by pidgeon777** [comment]
...but by commenting/removing that line, then one of my LSP servers crashes when looking for references with `navigator.lua`.

Only with `navigator.lua`, not when using `Telescope` or `Trouble` when looking for references.

I'm trying to investigate.

**Comment by pidgeon777** [comment]
When commenting/removing that line, the following params are sent to the LSP server:

```json
{ ["position"] = { ["line"] = 385,["character"] = 20,} ,["textDocument"] = { ["uri"] = file:///C:/I2C_MASTER.vhd,} ,} 
```

When leaving that line enabled, the following is sent:

```json
{ ["context"] = { ["includeDeclaration"] = false,} ,["position"] = { ["line"] = 385,["character"] = 16,} ,["textDocument"] = { ["uri"] = file:///C:/I2C_MASTER.vhd,} ,} 
```

I've tried to understand why the first request, subsequent to commenting/removing the line, would make one of my LSP servers crash.

The answer is here:

```log
[ERROR][2022-04-05 12:10:04] ...\vim\lsp\rpc.lua:420	"rpc"	"vhdl_ls"	"stderr"	"thread 'main' panicked at 'Invalid request\nMethod: textDocument/references\n error: missing field `context`', C:\\Users\\runneradmin\\.cargo\\registry\\src\\github.com-1ecc6299db9ec823\\lsp-server-0.3.4\\src\\msg.rs:163:17\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n"
```

The LSP server which crashes, expects to receive the `context` keyword when looking for references, but by commenting/removing the line I reported, this doesn't happen.

In short, this should be modified in `navigator.lua`: when looking for references, the `context` parameter should not be sent in the `textDocument/definitions` request (which is done to collect further references by your plugin) because it also seems not to be supported in the Microsoft LSP Protocol, but it must be sent when requesting the `textDocument/references` method only.

**Comment by pidgeon777** [comment]
This should fix the problem:

```lua
local async_ref = function()
  local ref_params = vim.lsp.util.make_position_params()
  local results = {}
  lsp.call_async('textDocument/definition', ref_params, function(err, result, ctx, config)
    trace(err, result, ctx, config)
    if result ~= nil then
      for i = 1, #result do
        if result[i].range == nil and result[i].targetRange then
          result[i].range = result[i].targetRange
        end
      end
    end
    results.definitions = { error = err, result = result, ctx = ctx, config = config }
    log(result)
    ctx = ctx or {}
    ctx.results = results
    ctx.combine = true
    ref_view(err, result, ctx, config)
  end) -- return asyncresult, canceller
  ref_params.context = { includeDeclaration = false }
  lsp.call_async('textDocument/references', ref_params, function(err, result, ctx, config)
    trace(err, result, ctx, config)
    results.references = { error = err, result = result, ctx = ctx, config = config }
    ctx = ctx or {}
    ctx.results = results
    ctx.combine = true
    ref_view(err, result, ctx, config)
  end) -- return asyncresult, canceller
end
```

But again, I'm not sure if this would be the best way to do it.

**Comment by pidgeon777** [comment]
My proposal above also includes the possible fix for this:

https://github.com/ray-x/navigator.lua/issues/163

**Comment by pidgeon777** [comment]
There is a remaining fix to add to close this issue.

This line:

https://github.com/ray-x/navigator.lua/blob/8b43ed23e254f7401adcedcd25f4b31f1af84fb2/lua/navigator/reference.lua#L153

must be removed and placed right above this line, instead:

https://github.com/ray-x/navigator.lua/blob/8b43ed23e254f7401adcedcd25f4b31f1af84fb2/lua/navigator/reference.lua#L172

**Comment by ray-x** [comment]
@pidgeon777 , can you confirm if the issue was fixed? thanks!
